File: frontend/src/types/transaction.ts
export type TransactionStatus = "CLEARED" | "HELD" | "APPROVED" | "REJECTED" | "ESCALATED";
export type RiskLevel = "LOW" | "MEDIUM" | "HIGH";

export interface Transaction {
  id: string;
  account_number: string;
  account_holder_name: string;
  amount: string;
  merchant_name: string;
  merchant_category: string | null;
  transaction_type: string;
  location_city: string;
  location_country: string;
  latitude: number | null;
  longitude: number | null;
  timestamp: string;
  status: TransactionStatus;
  risk_level: RiskLevel;
  risk_score: number;
  fraud_flags: string[];
  reviewed_by: string | null;
  reviewed_at: string | null;
  review_notes: string | null;
  created_at: string;
  updated_at: string | null;
}

export interface TransactionListResponse {
  transactions: Transaction[];
  total: number;
  limit: number;
  offset: number;
}

export interface TransactionFilters {
  status?: TransactionStatus;
  risk_level?: RiskLevel;
  account_number?: string;
  merchant_name?: string;
  date_from?: string;
  date_to?: string;
  min_amount?: number;
  max_amount?: number;
  limit?: number;
  offset?: number;
}

export interface ReviewRequest {
  decision: "APPROVED" | "REJECTED" | "ESCALATED";
  notes: string;
  reviewed_by: string;
}

export interface AccountStats {
  average_amount: string;
  transaction_count: number;
  common_locations: string[];
  first_transaction_date: string | null;
}

export interface AccountHistoryResponse {
  account_number: string;
  account_holder_name: string;
  contextual_transactions: Transaction[];
  recent_transactions: Transaction[];
  reviewed_transaction_time: string;
  stats: AccountStats;
}

export interface DashboardStats {
  held_count: number;
  approved_today: number;
  rejected_today: number;
  escalated_count: number;
  avg_review_time_minutes: number;
  transactions_by_risk: Record<string, number>;
}

File: backend/app/schemas/transaction.py
from datetime import datetime
from typing import Optional, List
from decimal import Decimal
from pydantic import BaseModel, Field
from app.models.transaction import TransactionStatus, RiskLevel


class TransactionBase(BaseModel):
    account_number: str
    account_holder_name: str
    amount: Decimal
    merchant_name: str
    merchant_category: Optional[str] = None
    transaction_type: str
    location_city: str
    location_country: str
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    timestamp: datetime


class TransactionCreate(TransactionBase):
    pass


class TransactionResponse(TransactionBase):
    id: str
    status: TransactionStatus
    risk_level: RiskLevel
    risk_score: int
    fraud_flags: List[str]
    reviewed_by: Optional[str] = None
    reviewed_at: Optional[datetime] = None
    review_notes: Optional[str] = None
    created_at: datetime
    updated_at: Optional[datetime] = None
    
    class Config:
        from_attributes = True


class TransactionListResponse(BaseModel):
    transactions: List[TransactionResponse]
    total: int
    limit: int
    offset: int


class ReviewRequest(BaseModel):
    decision: TransactionStatus = Field(..., description="Must be APPROVED, REJECTED, or ESCALATED")
    notes: str = Field(..., min_length=1, description="Review notes are required")
    reviewed_by: str = Field(..., min_length=1, description="Analyst name is required")


class AccountStats(BaseModel):
    average_amount: Decimal
    transaction_count: int
    common_locations: List[str]
    first_transaction_date: Optional[datetime]


class AccountHistoryResponse(BaseModel):
    account_number: str
    account_holder_name: str
    contextual_transactions: List[TransactionResponse]
    recent_transactions: List[TransactionResponse]
    reviewed_transaction_time: datetime
    stats: AccountStats


class StatsResponse(BaseModel):
    held_count: int
    approved_today: int
    rejected_today: int
    escalated_count: int
    avg_review_time_minutes: float
    transactions_by_risk: dict
